Move grid power and current metrics to Grid

daniel-zullo-frequenz opened the pull request:
Fixes #729

daniel-zullo-frequenz committed:
Add formula engine pool to grid

The formula engine pool is needed to generate grid metrics.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Initialize grid in data pipeline

The grid is initialized in the data pipeline, so that it can
instantiate the formula engine pool to generate grid metrics.
This caused circular imports, which is resolved by enclosing
the type hints in a `TYPE_CHECKING` block.

Also the component metadata is now retrieved from the component
graph, instead of the microgrid API client. This is because
asynchronous calls to the API are not suitable in the data
pipeline as this is a high level user facing interface.

The grid initialize() function needed to be changed to process
the metadata as a dictionary rather than as a ComponentMetadata,
and this is due to the fact that the component graph does not
effectively retrieve the metadata as a ComponentMetadata object.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Amend grid connection tests

All the grid connection tests now use the microgrid to instantiate
the grid connection.

The test that was checking for multiple grid connections was
removed, because the component graph does not accept multiple
grid connections.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Move grid power and current formulas to Grid

Now that there is a microgrid.grid object,
the grid metrics are published there.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update release notes

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented on RELEASE_NOTES.md:
```diff
@@ -10,7 +10,7 @@ This version ships an experimental version of the **Power Manager**, adds prelim
 - `microgrid.grid()`
 
-  * Similar to `microgrid.battery_pool()`, the Grid is now similarily accessed.
+  * Similar to `microgrid.battery_pool()`, the Grid is now similarly accessed.
```
I know it is not added by this pr, but is it me or does this sound pretty weird? We can change it in another PR cleaning up the release notes.

daniel-zullo-frequenz replied:
Yes, it does sound weird. I'll amend it and actually move it together with this PR release-notes changes because they are related

shsms replied:
Yes,  one of them has to be absolute and the other can be similar.  otherwise it is a circular dependency.

daniel-zullo-frequenz replied:
Rephrased and grouped together with this PR release-notes changes

llucax reviewed (commented):
LGTM so far!

daniel-zullo-frequenz commented on docs/tutorials/getting_started.md:
```diff
@@ -117,7 +117,7 @@ async def run() -> None:
 
     # Define your application logic here
-    grid_meter = microgrid.logical_meter().grid_power.new_receiver()
+    grid_meter = microgrid.grid().power.new_receiver()
```
@frequenz-floss/python-sdk-team the example here does not consider that `microgrid.grid()` can return `None`, should I amend the example? Or is it up to the user to check the documentation and add checks or error-handling as this is just an example?

shsms replied:
oh really?  I thought the stream at least should always be accessible.  Has nothing to do with whether metadata is available or not.  No grid metadata doesn't mean there's no grid component in the component graph, right?

And even if there's no grid component, the formula should still be available, so that people are able to build one general purpose algorithm, rather than one algorithm when grid power is available, one when it isn't, etc.

shsms replied:
In other words, I don't think we should put the grid formula into a class that can be None.  If we are doing this, then I think `grid` has to always be there.

daniel-zullo-frequenz replied:
Then I'll need to work out the Grid interface, I think I can make the `Fuse` attribute optional instead of the `Grid` object so that the microgrids configured as islands won't have a grid connection (fuse == None). Another alternative would be to use another class

llucax replied:
Yeah, good point. As we mentioned earlier, if the idea is to abstract away when we are on an island, we might want to still have a (virtual?) fuse with a current rate of `0`. So you can treat the grid connection generically.

llucax replied:
But yeah, I can see how makes us re-think the grid connection interface, but I think it is going in a good direction.

shsms replied:
> we might want to still have a (virtual?) fuse with a current rate of 0

true, maybe worth making an issue, or picking up immediately after this.

daniel-zullo-frequenz replied:
Added the commit _Refactor Grid interface_ but I can move it to a new PR if you think is better

shsms replied:
Sorry, I'm realizing it might be more complicated than that.  Or maybe I need to look into the current implementation more.  We always want to have a grid interface.  But not sure we'll always want to have a non-None fuse interface.  This is because the api might not give us fuse info,  but there would still be a grid component in the component graph.  

In that case,  using 0 value for fuse would be inaccurate, we just don't know the fuse value.  So we have to say it is None.

So there can be situations where fuse has to be None,  but will confirm on Monday.

shsms replied:
Even otherwise, the way to confirm that it is an island config, is by looking at the component graph, and not from the metadata.

shsms replied:
Is why dealing with it separately might be easier.  Although I might be completely wrong, haven't looked at the code yet.

llucax replied:
So no grid connection -> `fuse = 0`
Unknown fuse -> `fuse = None`, which in practice means infinite? Or what else can be done if we know there is a grid connection but have no fuse info or there is no fuse?

shsms replied:
No, it just means we don't know the fuse limit, and can't run actors that depend on the fuse limit, like a fuse safety actor.  It looks like this part can't be abstracted away, the way it stands.  Alternatively we can go to the edge team and ask them to confirm if we can always have this, because I don't see why not.  Fuse limit for a location can't be a hard thing to find out.

llucax replied:
Still, thinking about failure scenarios, if for some reason we don't receive a fuse (a bug in the microgrid service), we should still be able to operate, right? Only a few pieces might depend on the fuse info. So I think we should support that scenario, even if this force users to check if `fuse is not None` when then need to access the fuse info.

shsms replied:
> we should still be able to operate, right?

Yes, a lot of our actors don't depend on the fuse value, and I was not suggesting that we stop operating at all, just that we set `None` values for fuse, rather than 0.

And as you said, if an actor wants to explicitly protect against blowing the fuse, it can't.  In that scenario, it is up to the users to decide if they want to continue with the other actors anyway or not.

llucax replied:
> Yes, a lot of our actors don't depend on the fuse value, and I was not suggesting that we stop operating at all, just that we set `None` values for fuse, rather than 0.

Oh, OK, then I misunderstood your "No", I guess it was only for the how to interpret the `None`, not the `None` itself :+1:

daniel-zullo-frequenz replied:
The current source code on `v1.x.x` indicates that if there is grid connection then metadata and fuse must be available, otherwise it is a runtime error and the grid won't be initialized https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/src/frequenz/sdk/timeseries/grid.py#L60-L66

@tiyash-basu-frequenz can you please confirm that the metadata and fuse will be always set where there is a grid connection point? And is it a possibility that the fuse max current is unknown?

llucax replied:
Again, I wouldn't make the SDK depend on what the API currently does. Despite what we currently do, I think we should support cases where there is a grid connection but no fuse data for some reason. Just refuse to initialize the SDK because we don't have fuse info seems overkill.

tiyash-basu-frequenz replied:
I agree with Luca. The SDK should just look at the API, and should not have any dependency on the service implementation.

Regarding fuses, it's actually very difficult to say. If a microgrid owner configures one, the API will expose it. Otherwise not.

tiyash-basu-frequenz replied:
>Alternatively we can go to the edge team and ask them to confirm if we can always have this, because I don't see why not. Fuse limit for a location can't be a hard thing to find out.

We are not in control of this information. If a user provides this, we get it, otherwise we do not. If this blocks apps from running, then the app should issue an error, which will (in the future) get shown to the user in the UI.

daniel-zullo-frequenz replied:
Thanks! Then I'll remove the original/initial fuse checks allowing it to be `None` and clarify that fuse could be `None` for either islanded microgrids or it is just unknown for any reason

daniel-zullo-frequenz replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -23,6 +23,7 @@
 from ..timeseries._grid_frequency import GridFrequency
```
@frequenz-floss/python-sdk-team this naming `GridFrequency` is confusing to me as it somehow clashes with the grid connection `Grid`. Do you think the name is right or is it just me? I think a name around `MicrogridFrequency`  or `ComponentFrequency` would be more suitable. If we need to change it should be done in a separate PR. Please let me know your thoughts, thanks!

shsms replied:
I agree, `MicrogridFrequency` sounds like a better fit, based on our earlier discussions, by maybe even just `ACFrequency`, because there's only one value for one microgrid, whether it is connected to the grid or not.

llucax replied:
I would call it `MicrogridFrequency` for more context, `ACFrequency` sounds too vague, frequency where? Is it just a metric?

daniel-zullo-frequenz replied:
Yes, it is just a metric

daniel-zullo-frequenz replied:
I created the ticket https://github.com/frequenz-floss/frequenz-sdk-python/issues/770 to address this suggestion

daniel-zullo-frequenz replied:
or It could also be moved to `microgrid.grid`

llucax replied:
I don't think it is the best idea, because islands still have a frequency, and even if we will have a `grid` component for islands, it is weird to talk about a frequency for something where there is no current flowing.

daniel-zullo-frequenz replied:
But I think the same would apply for the metric `current()`. I understood we are now defining `Grid` as an abstract component regardless if there is a grid connection or not

shsms replied:
I would have liked that, but is different, because when there's no grid connection, we report 0 power and current values from `grid.current()`, etc,  but not so with frequency.  `microgrid.frequency()` feels like the right place for it.

daniel-zullo-frequenz replied:
Got it, then I'll keep the ticket with the initial suggestion just to rename it. Thanks for the clarification!

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/grid.py:
```diff
@@ -6,32 +6,135 @@
+
+from ..microgrid.component import Component, ComponentMetadata
 from ..microgrid.component._component import ComponentCategory
 from ..microgrid.fuse import Fuse
```
@frequenz-floss/python-sdk-team `Fuse` is defined in `frequenz.sdk.microgrid` while `Grid` is defined in `frequenz.sdk.timeseries`. Is this correct? just checking ...

shsms replied:
Probably not, nicer to keep them together in the timeseries package.

daniel-zullo-frequenz replied:
It will be addressed in https://github.com/frequenz-floss/frequenz-sdk-python/issues/771

daniel-zullo-frequenz committed:
Refactor Grid interface

Previously, the Grid interface returned `None` if the microgrid was
configured as an island.
This update ensures that the interface consistently returns a Grid
instance.
This change is necessary to treat the grid connection uniformly within
the microgrid, regardless of whether the microgrid is configured as
an island or not. Moreover, it guarantees that the `microgrid.grid()`
function always yields a Grid instance for metric retrieval.

Now the rated current of the fuse is set to zero in case of an
islanded microgrid.
And the fuse of the grid connection point is now set to `None` when
the grid connection component metadata lacks information about the
fuse. Thus, a fuse set to `None` won't prevent the initialization
of the Grid instance.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Amend docstring for Grid.current property

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented:
Is this ready for review again?

daniel-zullo-frequenz commented:
> Is this ready for review again?

Yes, it is. Somehow the re-request review is not available for you and Sahas.

llucax commented on RELEASE_NOTES.md:
```diff
@@ -8,6 +8,27 @@
 
+- `microgrid.grid()`
+  - A `Grid` object will always be instantiated now, even if the microgrid is not connected to the grid (islanded microgrid).
+  - The fuse will be set to `None` to indicate that the grid connection point is not available, either because the fuse is unknown (not configured) or because the microgrid is islanded.
```
I'm not sure if I'm forgetting something or if there is a technical difficulty to implement it differently, but wouldn't it make more sense to still set a fuse with a rate of 0 for islands, as there could never be any electricity flow through the grid connection in this case?

daniel-zullo-frequenz replied:
I understood `Grid` should be agnostic to the "type" of the microgrid and shouldn't make any difference whether the microgrid is islanded or not. The user can always check the component graph if needs to know there is a `grid` component or not. I tried to state the most known examples in the release notes but it seems this is misleading in the sense that makes you think the Grid.fuse can be used to somehow classify the microgrid

llucax replied:
> I understood Grid should be agnostic to the "type" of the microgrid and shouldn't make any difference whether the microgrid is islanded or not.

Yes, me too, I think we are on the same page with this one.

> The user can always check the component graph if needs to know there is a grid component or not.

Also agreed.

> I tried to state the most known examples in the release notes but it seems this is misleading in the sense that makes you think the Grid.fuse can be used to somehow classify the microgrid

I think there are 2 issues here:

1. The release note is misleading, specifically "The fuse will be set to `None` to indicate that the grid connection point is not available". This is not true, right? If `Fuse` is `None` is because the `Fuse` component / metadata (not sure if it came from the metadata or not) is not present in the microgrid API response, correct?

2. For applications to be able to use the `.grid()` in the most general way, I think it would make sense to create a fuse with a rate of 0 for islands instead of leaving the `Fuse` as `None`. This is because if we leave it at `None`, then users needs to do a special case for unknown fuse state, so they will probably need to check the component graph to see if there is a `Grid` component or not, and then if there is not, say "OK, then no current can flow through the grid" which is effectively the same as having a fuse of 0, no electricity can flow (or the fuse will be blown). So any code dealing with **existing** fuses, will behave correctly for islands without having to go check out if we are on an island or not.

I think it might be even good to have a `is_connected()` method in `Grid` as a shortcut to find out if we are connected to the grid or not (but that's for another PR/time).

daniel-zullo-frequenz replied:
> If Fuse is None is because the Fuse component / metadata (not sure if it came from the metadata or not) is not present in the microgrid API response, correct?

That's right

> For applications to be able to use the .grid() in the most general way, I think it would make sense to create a fuse with a rate of 0 for islands instead of leaving the Fuse as None. This is because if we leave it at None, then users needs to do a special case for unknown fuse state, so they will probably need to check the component graph to see if there is a Grid component or not, and then if there is not, say "OK, then no current can flow through the grid" which is effectively the same as having a fuse of 0, no electricity can flow (or the fuse will be blown). So any code dealing with existing fuses, will behave correctly for islands without having to go check out if we are on an island or not.

These are good points and make sense to me. Somehow I got confused when we previously discussed about it. I'll amend the PR accordingly.

> I think it might be even good to have a is_connected() method in Grid as a shortcut to find out if we are connected to the grid or not (but that's for another PR/time).

I'll create an issue for this

daniel-zullo-frequenz replied:
Updated to set the rated current of the grid fuse to `Current.zero` in case of a microgrid configured as an island.
Also updated the release notes.

daniel-zullo-frequenz replied:
Also created https://github.com/frequenz-floss/frequenz-sdk-python/issues/787

llucax replied:
> These are good points and make sense to me. Somehow I got confused when we previously discussed about it. I'll amend the PR accordingly.

OK, @shsms @matthias-wende-frequenz  FYI, just in case I misunderstood anything myself. It would be good to confirm we all think this is the best approach.

shsms replied:
This looks like the right approach, saying `None` only when there insufficient info,  and we're not even having to do a separate lookup of the component graph, so that's great as well!

llucax reviewed (commented):
Just one small comment about the fuse in islands.

shsms commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -199,7 +201,18 @@ def grid(
+                component_category={ComponentCategory.GRID},
+            )
+            initialize_grid(
+                components=components,
```
Just one comment: I see that this `components` param was already here in this form, but could you remove it, and do the `component_graph` access inside the `initialize_grid` method?

That would make it so much easier to read, and remove confusing docstring for the argument, etc.

daniel-zullo-frequenz replied:
Sure, that makes sense. On top of your points this was unnecessarily  filtering twice the same grid components.
Updated as suggested and removed a no longer needed test as it was directly calling `initialize()` by passing the component graph which does not allow to configure more than 1 grid component

llucax commented on src/frequenz/sdk/timeseries/grid.py:
```diff
@@ -78,7 +81,17 @@ def initialize(
             raise RuntimeError("Grid metadata is None")
 
-        fuse = grid_connections[0].metadata.fuse
+        # The component graph does not store the metadata as a
+        # ComponentMetadata object but as a dict and the type-checker believes
+        # it is a ComponentMetadata object.
+        metadata = grid_connections[0].metadata
```
This is weird, why is the type wrong in the component graph? I mean why the type-checker believes it is a `ComponentMetadata` object if it's not?

daniel-zullo-frequenz replied:
The method `components()` in the Component Graph returns a `set[Component]` where each `Component` is made of `component_id: int`, `category: ComponentCategory`, `type: ComponentType` and `metadata: ComponentMetadata`

However a `dict` type for `metadata` field is returned when calling `components()` for the grid component. So it seems to me the type-checker is not working correctly and the implementation in Component Graph might not be right either. Please check the code below for further details

```python
    fuse = Fuse(max_current=Current.from_amperes(123.0))

    components = {
        Component(1, ComponentCategory.GRID, None, GridMetadata(fuse)),
        Component(2, ComponentCategory.METER),
    }
    connections = {
        Connection(1, 2),
    }

    # pylint: disable=protected-access
    graph = gr._MicrogridComponentGraph(components=components, connections=connections)

    components = graph.components()
    assert len(components) == 2
    for component in components:
        if component.component_id == 1:
            assert component.category == ComponentCategory.GRID
            assert component.metadata is not None
            assert isinstance(component.metadata, dict)  # PASSED
            assert isinstance(component.metadata, GridMetadata)  # FAILED
```

Moreover, this was kind of considered expected behaviour according to existent component graph tests, see https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/tests/microgrid/test_graph.py#L836-L855

llucax replied:
OK, I see this problem is well beyond this PR.

daniel-zullo-frequenz replied:
Created the ticket to fix it https://github.com/frequenz-floss/frequenz-sdk-python/issues/789

llucax commented on src/frequenz/sdk/timeseries/grid.py:
```diff
@@ -78,7 +81,17 @@ def initialize(
+        metadata = grid_connections[0].metadata
+
+        fuse: Fuse | None = None
+        if isinstance(metadata, ComponentMetadata):
+            fuse = metadata.fuse
+        elif isinstance(metadata, dict):
+            fuse_dict = metadata.get("fuse", None)
+            fuse = Fuse(**fuse_dict) if fuse_dict else None
```
Also, can it really happen that it's either of both and not always a `dict` then? Why do we have the `if isinstance(metadata, ComponentMetadata)` at all?

daniel-zullo-frequenz replied:
Previous to the latest suggestion from Sahas, the `initialize()` method accepted a `Iterable[Component]` type as input parameter. The data pipeline was passing the component retrieved from the Component Graph but in a test it was just passing `Components` without using the component graph. As I have explained in https://github.com/frequenz-floss/frequenz-sdk-python/pull/758/files#r1401933878, the component graph does not retrieve the right object type and the type-checker failed to validate this.

In the current version, we can remove checking for `ComponentMetadata` as we will be only getting the components from the Component Graph but this is hack caused by the Component Graph...ideally we should only be checking for `metadata is not None`

shsms replied:
yes, this is definitely wrong, the component graph shouldn't leak details about its internal implementation (of converting to dict and back), especially in this buggy fashion, that we have to deal with in other parts of the code.

llucax replied:
Alright, so I'd say removed the unnecessary check in this PR and create a new issue to fix the type-leakage in the component graph.

daniel-zullo-frequenz replied:
Updated, the diff is here https://github.com/frequenz-floss/frequenz-sdk-python/compare/66abc038b0b9896a993c2b0b7000c43242526455..c10d6e2ced5491a112f29a9f6965de13e7b9b0ed

llucax reviewed (commented):
Just one question about typing confusion, but otherwise LGTM!

daniel-zullo-frequenz commented:
Rebased onto latest v1.x.x
